### ISSUE TITLE
Enable a mode where a txn-commits wait for on-disk-persistence

### ISSUE THREAD

**Issue by shriraml-ms** [microsoft/CCF#5093]
**Is your feature request related to a problem? Please describe.**
Yes, currently a txn can be globally committed without having the writes hit the disk because the commit doesn't wait for writes to be flushed to disk. This means that during DR, a previously globally committed txn can be lost. In systems that require higher guarantees on durability, it would greatly help if the ccf-commit flushed and waited for the writes to hit the disk.
Additional comments:
- This feature would obviously come with some perf-penalty. It would be good to have a perf test that shows comparative perf & have the mode/flag come with a pointer to documentation that lists the approx perf penalty. This would allow users to make the right tradeoff for their app.
- In cases where the host storage is malicious or misbehaving, the benefits to durability from this feature are obviously questionable - this is something that needs to be documented as well.
- I'm not sure if introducing an untrusted entity (host/storage-driver) into the critical path of commit, introduces any other attack vectors - this likely needs to be understood as well.

**Describe the solution you'd like**
Add a configuration switch that tells CCF to flush and wait for writes to hit the disk before marking a txn as globally committed.

**Describe alternatives you've considered**
One alternative is to run CCF across many more nodes but this can get expensive quickly.

**Additional context**
n/a

**Comment by achamayou** [comment]
Thank @shriraml-ms for opening this. There are a few options for "wait for the write to hit the disk", with different tradeoffs:

1. In consensus, nodes only ack an entry once their respective host tells them it has been sync'ed to disk. This is building disk persistence into consensus watermark, provides the strongest durability probability (*) but also the highest performance cost. 
2. In consensus, the primary builds in its local sync watermark into the commit level it establishes. This is weaker, but likely less expensive.
3. Consensus keeps track of sync watermark across the network, but does not gate commit on it. If a receipt is requested however, it is only emitted once the sync watermark has passed the transaction id. Weaker still, commit retains current attestation-based semantics, but receipts (signed commit) come with durability best effort across the network.
4. Consensus keeps track of sync watermark on the local node emitting the receipt (not necessarily the primary). Weak, commit retains existing attestation-based semantics, but receipts (opposable commit) come with durability effort on one node.

2. and 4. sound a bit odd for CCF, but are probably the closest to existing DBMS, which typically only consider local disk sync, and tend to have a separate, asynchronous remote replication mechanism.

(*) using durability probability rather than guarantee, because the host is not attested and under control of the attacker in our threat model. We tend to use guarantee for properties that hold under our threat model.

To your point:

> One alternative is to run CCF across many more nodes but this can get expensive quickly.

This depends on the relative failure probabilities of nodes vs local persistence, and their respective costs. As throughputs get higher, storage costs tend to increase faster than linearly (eg. E80 are twice as expensive as E70, despite offering 25% more throughput). Systems like SAP HANA for example tend to rely primarily on replication for that reason.

**Comment by shriraml-ms** [comment]
thank you @achamayou. 
- Nodes ack'ing an entry on after disk-flush is definitely the strongest.
- Nodes ack'ing without waiting for disk-flush while the current primary waits for local-disk-flush is definitely weaker. I'll also add that I'm not sure it is adding any value (compared to not waiting for local disk flush). In a DR scenario, how do you find the most recent ledger? You can't trust any of the other nodes that weren't primary (because they may not have received the entries) ... but you can't trust the primary either because that introduces a single point of compromise where a malicious storage host can tamper with the ledger. IMO, this may not be a useful option to expose. I'm curious to hear your thoughts.
- Requiring a request of receipt to do on-disk-persistence could be an option... but that is clubbing two things into one - what if an app wants "disk-level-commit" but doesn't need receipts for every txn. That app wouldn't have a way to request just the commits without paying the extra cost for receipt generation. IMO, it would keep things flexible if receipts and commits were kept separate (& apps can choose the desired behavior).
- Agree that 4 & 5 sound a bit odd - not worth pursuing IMO.

Thank you.

**Comment by achamayou** [comment]
> You can't trust any of the other nodes that weren't primary (because they may not have received the entries) ... but you can't trust the primary either because that introduces a single point of compromise where a malicious storage host can tamper with the ledger.

A malicious primary host can delete data, but not tamper with it, since it is protected by signatures and tags produced by keys that are kept in enclave.

In our threat model, all hosts are untrusted. They could all be lying about fsync all the time, or even honestly reporting fsync completed while another process is actively deleting the files as soon as they are written. Waiting for fsync can never provide a security guarantee, and all tamper protection must instead rely on cryptographic constructs.

> IMO, this may not be a useful option to expose.

That seems to be default for replicated Postgres: fsync on primary and async replication. I'm not sure how to quantify useful here without user input.

https://www.postgresql.org/docs/current/runtime-config-wal.html
https://www.postgresql.org/docs/current/warm-standby.html#SYNCHRONOUS-REPLICATION

> what if an app wants "disk-level-commit" but doesn't need receipts for every txn.

Receipt generation is fairly inexpensive (a few in-memory reads for recent transactions), and not likely to be the bottleneck in this case. Without a receipt, a user does not have an opposable or transferrable proof of commit.

**Comment by shriraml-ms** [comment]
> A malicious primary host can delete data, but not tamper with it, since it is protected by signatures and tags produced by keys that are kept in enclave. 

Yes, this is exactly the problem - it can delete entries that were globally committed since the primary is the only replica that has the entries in its disk (in the scenario where only the primary flushes to disk before commit).

> In our threat model, all hosts are untrusted. They could all be lying about fsync all the time, or even honestly reporting fsync completed while another process is actively deleting the files as soon as they are written. Waiting for fsync can never provide a security guarantee, and all tamper protection must instead rely on cryptographic constructs.

The problem here is that the cryptographic constructs don't provide protection against deletion at the tail end (or most recent entries). IMO, CCF adds this protection (from tail-delete) by de-centralizing with consensus. In other words, if we require disk flushes on all replicas then to delete the tail-entries, at least F+1 hosts need to be compromised in the worst case & all nodes need to be compromised in the best case. And once the majority of the network is compromised then all bets are off anyway, right?

> That seems to be default for replicated Postgres: fsync on primary and async replication. I'm not sure how to quantify useful here without user input.

The typical Replicated Postgres configuration you are referring to is like an active-passive configuration & not a decentralized quorum-based configuration (which is what CCF is) - they are too different to compare directly. The answers to my original questions for that option (like - how you would find the longest ledger reliably during DR) would give you an idea of the usefulness of this option for CCF. And yes, further user input would also provide some clarity.

> Receipt generation is fairly inexpensive (a few in-memory reads for recent transactions), and not likely to be the bottleneck in this case. Without a receipt, a user does not have an opposable or transferrable proof of commit.
Fair enough. If we go down this option, then all the docs would need to clarify that 
 receipts are required to be generated for a truly global commit (& the definition of "global commit" would also need to be adjusted accordingly). 

I personally perfer to think of commits and receipts as being separate things (as an end user, I think of commit as a signal of achieving durability and receipts as cryptographic proof of that durability) but that is just my opinion... you could probably get more clarity from further user-input.
